Bluetooth: Mesh: Composition data page 0 traversal

trond-snekvik opened the pull request:
Adds a parsing mechanism for Composition data page 0 in the Config
Client API, and uses it in the shell module to parse the incoming
composition data.

Demonstrates this API in the mesh_provisioner sample, allowing it to provision the basic mesh sample, as suggested by @pabigot in #33476.

cc @tsvehagen

trond-snekvik committed:
Bluetooth: Mesh: Rename bt_mesh_cfg_comp_data_get's status param to rsp

The first byte of the composition data status message is the returned
page index, not the status of the request. This is now reflected in the
API.

Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>

trond-snekvik committed:
Bluetooth: Mesh: Composition data page 0 traversal

Adds a parsing mechanism for Composition data page 0 in the Config
Client API, and uses it in the shell module to parse the incoming
composition data.

Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>

trond-snekvik committed:
Bluetooth: Mesh: Make provisioner sample bind the AppKey to all models

Instead of only configuring the Health Server model of the provisioned
device, the Provisioner will instead walk the target's composition data,
and bind each model to the same AppKey. This allows the provisioner to
be used with the base mesh sample, and demonstrates the new composition
data page 0 API.

Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>

jhedberg commented on samples/bluetooth/mesh_provisioner/src/main.c:
```diff
@@ -138,34 +140,67 @@ static void configure_node(struct bt_mesh_cdb_node *node)
 	}
 
+	elem_addr = node->addr;
+	while (!bt_mesh_comp_elem_pull(&comp, &elem)) {
```
`== 0` would be clearer to not make the reader think the function returns bool.

jhedberg replied:
Alternatively, rethink the return value type (bool, enum, give back the elem pointer, etc)

trond-snekvik replied:
Now returning a pointer, to avoid returning bool to indicate errors.

joerchan commented on include/bluetooth/mesh/cfg_cli.h:
```diff
@@ -936,6 +940,97 @@ int32_t bt_mesh_cfg_cli_timeout_get(void);
+ *  @return The model ID of the vendor model at the given index.
+ */
+struct bt_mesh_mod_id_vnd
+bt_mesh_comp_elem_mod_vnd(struct bt_mesh_comp_elem *elem, int idx);
```
Dont' have to break this line perhaps?

jhedberg commented on subsys/bluetooth/mesh/shell.c:
```diff
@@ -574,81 +574,79 @@ static int cmd_ident(const struct shell *shell, size_t argc, char *argv[])
+	shell_print(shell, "\tFeatures 0x%04x", comp.feat);
 
-		shell_print(shell, "\tElement @ 0x%04x:", loc);
+	while (!bt_mesh_comp_elem_pull(&comp, &elem)) {
```
Same here

joerchan commented on include/bluetooth/mesh/cfg_cli.h:
```diff
@@ -936,6 +940,97 @@ int32_t bt_mesh_cfg_cli_timeout_get(void);
+int bt_mesh_comp_page_0_get(struct bt_mesh_comp_page_0 *comp,
+			    struct net_buf_simple *buf);
+
+/** @brief Pull an element from a composition data page.
```
from a composition data page 0?

More instances in this function API description.

If I'm following this right.

joerchan commented on include/bluetooth/mesh/cfg_cli.h:
```diff
@@ -936,6 +940,97 @@ int32_t bt_mesh_cfg_cli_timeout_get(void);
+};
+
+/** Composition data page 0 element representation */
+struct bt_mesh_comp_elem {
```
Would `bt_mesh_comp_page_0_elem` be more clear?

trond-snekvik replied:
Probably, yes. I renamed both the composition data page and the elem to use a `bt_mesh_comp_p0` prefix.

jhedberg commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -2064,3 +2067,61 @@ void bt_mesh_cfg_cli_timeout_set(int32_t timeout)
+	}
+
+	elem->_buf =
+		(uint8_t *)net_buf_simple_pull_mem(page->_buf, modlist_size);
```
Looks like an unnecessary type cast

joerchan commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -2064,3 +2067,61 @@ void bt_mesh_cfg_cli_timeout_set(int32_t timeout)
+	return 0;
+}
+
+int bt_mesh_comp_elem_pull(const struct bt_mesh_comp_page_0 *page,
```
similarly, would `bt_mesh_comp_page_0_elem_pull` be more clear?

jhedberg commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -2064,3 +2067,61 @@ void bt_mesh_cfg_cli_timeout_set(int32_t timeout)
+
+uint16_t bt_mesh_comp_elem_mod(struct bt_mesh_comp_elem *elem, int idx)
+{
+	return sys_get_le16(&elem->_buf[idx * 2]);
```
Should you check elem->nsig first?

trond-snekvik replied:
Added a bounds check, but there's no room for an error code in the return types of these functions, so I have made them return `0xffff`. This can also be a valid return value though, so I think it's the lesser of two evils. I'd prefer not to change the return type, as this would mess up the vnd function.

jhedberg commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -2064,3 +2067,61 @@ void bt_mesh_cfg_cli_timeout_set(int32_t timeout)
+struct bt_mesh_mod_id_vnd
+bt_mesh_comp_elem_mod_vnd(struct bt_mesh_comp_elem *elem, int idx)
+{
+	size_t offset = elem->nsig * 2 + idx * 4;
```
Same here, since this is a public API we might want to do a bounds check first?

trond-snekvik commented:
Marking with DNM until @pabigot's comments in #33476 are resolved

trond-snekvik committed:
Bluetooth: Mesh: Provisioner sample: Cooperative thread

Runs the Provisioner sample's main thread in a cooperative priority to
avoid calling the Bluetooth APIs from a preemptive thread.

Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>

trond-snekvik committed:
Bluetooth: Mesh: Provisioner sample: Blocking self-configuration

Adds blocking operation to the self configuration step by waiting for
the status response. This ensures that the provisioner is fully
configured, and reduces the number of loopback buffers required for the
self-configuring.

Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>

jhedberg commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -2064,3 +2067,68 @@ void bt_mesh_cfg_cli_timeout_set(int32_t timeout)
+
+uint16_t bt_mesh_comp_p0_elem_mod(struct bt_mesh_comp_p0_elem *elem, int idx)
+{
+	if (idx >= elem->nsig) {
+		return 0xffff;
+	}
```
Don't we want to take into use `CHECKIF()` from `include/sys/check.h`? I've seen that used in many Bluetooth audio related patches. It seems the general trend is to move toward that for all input parameter checking.

trond-snekvik replied:
I wasn't aware of this. Makes sense, I'll change it.

trond-snekvik committed:
Bluetooth: Mesh: Provisioner sample: Permit 16 nodes

Ups the CDB_NODE_COUNT config to fit 16 nodes, to make the sample more
usable for general mesh testing.

Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>

pabigot commented:
This needs to be merged to complete the resolution of #31031.

jhedberg commented on subsys/bluetooth/mesh/cfg_cli.c:
```diff
@@ -2064,3 +2068,68 @@ void bt_mesh_cfg_cli_timeout_set(int32_t timeout)
+int bt_mesh_comp_p0_get(struct bt_mesh_comp_p0 *page,
+			struct net_buf_simple *buf)
+{
+	if (buf->len < 10) {
+		return -EINVAL;
+	}
```
`CHECKIF()` here as well?

trond-snekvik replied:
I left it out on purpose, as it's doing a comparison on on-air data. With `CHECKIF`, it can turn into an assert or be skipped, which would force the application to do their own check. This would sort of negate the intended effect of this macro

jhedberg replied:
Right. Makes sense.
